daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,177 @@
+ *
+ * There are 2 main event handlers an implementing class has to worry about: one for Terms, and
+ * another for Types.  These are called as the DSOM is walked.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ */
+abstract class AbstractDSOMWalker {
+
+  /**
+   * Method to be called on the beginning of the traversal.  It is recommended to add some
+   * sort of wrapper element to a stack if you're doing a typical stack-based traversal.
+   *
+   * @param root the root element of the DFDL Schema
+   */
+  protected def onWalkBegin(root: RootView): Unit
+
+  /**
+   * Method to be called when the traversal concludes.  It is recommended to put any post-processing
+   * and anything to tidy up the stack or the result here.
+   *
+   * @param root the root element of the DFDL Schema
+   */
+  protected def onWalkEnd(root: RootView): Unit
+
+  /**
+   * Method to be called whenever any element that is a Term is encountered.
+   * This applies to Sequence, Choice, GroupRef, ElementBase, etc.
+   *
+   * It is highly recommended that, when implementing this method, you pattern match
+   * some of these different sub-types at some point to handle each accordingly.
+   *
+   * @param termElement the term element
+   */
+  def onTermBegin(termElement: TermView): Unit
+
+  /**
+   * Method to be called when a Term element has finished processing
+   *
+   * @param termElement the term element
+   */
+  def onTermEnd(termElement: TermView): Unit
+
+  /**
+   * Method to be called when either a Simple or Complex Type element has been encountered.
+   * This is just the element for <simpleType> or <complexType>, so no implementation is necessary if you
+   * do not care about these wrapper portion; its children will be walked automatically.
+   *
+   * @param typeElement either a Simple or Complex type
+   */
+  def onTypeBegin(typeElement: TypeView): Unit
+
+  /**
+   * Method to be called when either a Simple or Complex Type element has been finished
+   * processing.  See onTypeBegin method description
+   *
+   * @param typeElement either a Simple or Complex type
+   */
+  def onTypeEnd(typeElement: TypeView): Unit
+
+  /**
+   * Starts a DSOM walk from the Root object.  By default, the walk is completed recursively in the helper
+   * method below, but a custom walk can be defined by overriding
+   * this method instead and calling local event handlers.
+   * @param schemaSetRoot The root element of the DFDL Schema.  This will be the starting point of the traversal
+   */
+  def walkFromRoot(schemaSetRoot: RootView): Unit = {
+    onWalkBegin(schemaSetRoot)
+    walkerHelper(schemaSetRoot)
+    // onWalkEnd will return the value returned by this method!
```

Comment:
Can remove this comment now that walkerEnd doesn't return anything.

Reply:
Agreed - Resolved in the latest revision